src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.DeclarationAnalysisData.cs

```diff
@@ -4,51 +4,55 @@
 
 using System;
 using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.PooledObjects;
 
 namespace Microsoft.CodeAnalysis.Diagnostics
 {
     internal abstract partial class AnalyzerDriver
     {
-        internal sealed class DeclarationAnalysisData
+        internal readonly struct DeclarationAnalysisData : IDisposable
         {
-            public DeclarationAnalysisData(
-                SyntaxNode declaringReferenceSyntax,
-                SyntaxNode topmostNodeForAnalysis,
-                ImmutableArray<DeclarationInfo> declarationsInNodeBuilder,
-                ImmutableArray<SyntaxNode> descendantNodesToAnalyze,
-                bool isPartialAnalysis)
-            {
-                DeclaringReferenceSyntax = declaringReferenceSyntax;
-                TopmostNodeForAnalysis = topmostNodeForAnalysis;
-                DeclarationsInNode = declarationsInNodeBuilder;
-                DescendantNodesToAnalyze = descendantNodesToAnalyze;
-                IsPartialAnalysis = isPartialAnalysis;
-            }
-
             /// <summary>
             /// GetSyntax() for the given SyntaxReference.
             /// </summary>
-            public SyntaxNode DeclaringReferenceSyntax { get; }
+            public readonly SyntaxNode DeclaringReferenceSyntax;
 
             /// <summary>
             /// Topmost declaration node for analysis.
             /// </summary>
-            public SyntaxNode TopmostNodeForAnalysis { get; }
+            public readonly SyntaxNode TopmostNodeForAnalysis;
 
             /// <summary>
             /// All member declarations within the declaration.
             /// </summary>
-            public ImmutableArray<DeclarationInfo> DeclarationsInNode { get; }
+            public readonly ImmutableArray<DeclarationInfo> DeclarationsInNode;
 
             /// <summary>
             /// All descendant nodes for syntax node actions.
             /// </summary>
-            public ImmutableArray<SyntaxNode> DescendantNodesToAnalyze { get; }
+            public readonly ArrayBuilder<SyntaxNode> DescendantNodesToAnalyze = ArrayBuilder<SyntaxNode>.GetInstance();
```

Comment:
Should this type receive a fully-formed Frozen builder instead?

Reply:
that seems like a good idea.  i'll change to that :)

Reply:
Ah no.  this doesn't work.  this eventually gets passed to a method that operates over an IEnumerable.  That entrypoint is given lots of different types (And i don't want to box the FrozenArrayBuilder into the IEnumerable. s o i'll leave as is.